File: src/providers/ollama.ts
/**
 * Ollama provider — local LLM inference via Ollama REST API.
 */

import type { Message, ToolCall } from "../types/message.js";
import type { StreamEvent, ToolCallComplete } from "../types/events.js";
import { createAssistantMessage } from "../types/message.js";
import type { Provider, APIToolDef, ModelInfo, ProviderConfig } from "./base.js";

export class OllamaProvider implements Provider {
  readonly name = "ollama";
  private baseUrl: string;
  private defaultModel: string;

  constructor(config: ProviderConfig) {
    this.baseUrl = (config.baseUrl ?? "http://localhost:11434").replace(/\/$/, "");
    this.defaultModel = config.defaultModel ?? "llama3.1";
  }

  private convertMessages(
    messages: Message[],
    systemPrompt: string,
  ): unknown[] {
    const converted: unknown[] = [];
    if (systemPrompt) {
      converted.push({ role: "system", content: systemPrompt });
    }
    for (const msg of messages) {
      if (msg.role === "system") continue;

      if (msg.role === "assistant" && msg.toolCalls?.length) {
        converted.push({
          role: "assistant",
          content: msg.content || "",
          tool_calls: msg.toolCalls.map((tc) => ({
            id: tc.id,
            type: "function",
            function: {
              name: tc.toolName,
              arguments: JSON.stringify(tc.arguments),
            },
          })),
        });
      } else if (msg.role === "tool" && msg.toolResults?.length) {
        for (const tr of msg.toolResults) {
          converted.push({
            role: "tool",
            content: tr.output,
            tool_call_id: tr.callId,
          });
        }
      } else {
        converted.push({
          role: msg.role === "user" ? "user" : msg.role === "assistant" ? "assistant" : msg.role,
          content: msg.content,
        });
      }
    }
    return converted;
  }

  private convertTools(tools?: APIToolDef[]): unknown[] | undefined {
    if (!tools?.length) return undefined;
    return tools.map((t) => ({
      type: "function",
      function: {
        name: t.function.name,
        description: t.function.description,
        parameters: t.function.parameters,
      },
    }));
  }

  async *stream(
    messages: Message[],
    systemPrompt: string,
    tools?: APIToolDef[],
    model?: string,
  ): AsyncGenerator<StreamEvent, void> {
    const m = model ?? this.defaultModel;
    const msgs = this.convertMessages(messages, systemPrompt);
    const body: Record<string, unknown> = {
      model: m,
      messages: msgs,
      stream: true,
    };
    const ollamaTools = this.convertTools(tools);
    if (ollamaTools) body.tools = ollamaTools;

    let res: Response;
    try {
      res = await fetch(`${this.baseUrl}/api/chat`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      });
    } catch (err) {
      yield { type: "error", message: `Ollama request failed: ${err}` };
      return;
    }

    if (!res.ok) {
      yield { type: "error", message: `Ollama HTTP ${res.status}: ${await res.text()}` };
      return;
    }

    const reader = res.body?.getReader();
    if (!reader) {
      yield { type: "error", message: "Ollama: response body is not readable" };
      return;
    }

    const decoder = new TextDecoder();
    let buffer = "";
    let inThinkBlock = false;
    let thinkBuffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        if (!line.trim()) continue;
        let chunk: any;
        try {
          chunk = JSON.parse(line);
        } catch {
          continue;
        }

        // Handle Ollama errors embedded in the stream
        if (chunk.error) {
          yield { type: "error", message: `Ollama: ${chunk.error}` };
          return;
        }

        if (chunk.message?.content) {
          // Parse <think> tags — yield thinking_delta for think blocks, text_delta for normal text
          let remaining = chunk.message.content;
          while (remaining.length > 0) {
            if (inThinkBlock) {
              const closeIdx = remaining.indexOf("</think>");
              if (closeIdx !== -1) {
                thinkBuffer += remaining.slice(0, closeIdx);
                if (thinkBuffer) {
                  yield { type: "thinking_delta" as const, content: thinkBuffer };
                }
                thinkBuffer = "";
                inThinkBlock = false;
                remaining = remaining.slice(closeIdx + "</think>".length);
              } else {
                thinkBuffer += remaining;
                remaining = "";
              }
            } else {
              const openIdx = remaining.indexOf("<think>");
              if (openIdx !== -1) {
                if (openIdx > 0) {
                  yield { type: "text_delta", content: remaining.slice(0, openIdx) };
                }
                inThinkBlock = true;
                remaining = remaining.slice(openIdx + "<think>".length);
              } else {
                yield { type: "text_delta", content: remaining };
                remaining = "";
              }
            }
          }
        }

        if (chunk.message?.tool_calls) {
          for (const tc of chunk.message.tool_calls) {
            const callId = tc.id ?? crypto.randomUUID();
            const toolName = tc.function?.name ?? "unknown";
            yield {
              type: "tool_call_start",
              toolName,
              callId,
            };
            const args =
              typeof tc.function?.arguments === "string"
                ? JSON.parse(tc.function.arguments)
                : tc.function?.arguments ?? {};
            yield {
              type: "tool_call_complete",
              callId,
              toolName,
              arguments: args,
            } satisfies ToolCallComplete;
          }
        }

        if (chunk.done) {
          const inputTokens = chunk.prompt_eval_count ?? 0;
          const outputTokens = chunk.eval_count ?? 0;
          yield {
            type: "cost_update",
            inputTokens,
            outputTokens,
            cost: 0,
            model: m,
          };
        }
      }
    }
  }

  async complete(
    messages: Message[],
    systemPrompt: string,
    tools?: APIToolDef[],
    model?: string,
  ): Promise<Message> {
    const m = model ?? this.defaultModel;
    const msgs = this.convertMessages(messages, systemPrompt);
    const body: Record<string, unknown> = {
      model: m,
      messages: msgs,
      stream: false,
    };
    const ollamaTools = this.convertTools(tools);
    if (ollamaTools) body.tools = ollamaTools;

    const res = await fetch(`${this.baseUrl}/api/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });

    if (!res.ok) {
      throw new Error(`Ollama HTTP ${res.status}: ${await res.text()}`);
    }

    const data: any = await res.json();
    const content = data.message?.content ?? "";
    let toolCalls: ToolCall[] | undefined;

    if (data.message?.tool_calls?.length) {
      toolCalls = data.message.tool_calls.map((tc: any) => ({
        id: tc.id ?? crypto.randomUUID(),
        toolName: tc.function?.name ?? "unknown",
        arguments:
          typeof tc.function?.arguments === "string"
            ? JSON.parse(tc.function.arguments)
            : tc.function?.arguments ?? {},
      }));
    }

    return createAssistantMessage(content, toolCalls);
  }

  listModels(): ModelInfo[] {
    return [];
  }

  async fetchModels(): Promise<ModelInfo[]> {
    try {
      const res = await fetch(`${this.baseUrl}/api/tags`);
      if (!res.ok) return [];
      const data: any = await res.json();
      return (data.models ?? []).map((m: any) => {
        // Detect vision support from model families
        // Presence of "clip" or "llava" in families indicates vision capability
        const families = m.details?.families ?? [];
        const supportsVision = Array.isArray(families) &&
          families.some((f: string) => f.includes("clip") || f.includes("llava"));

        return {
          id: m.name as string,
          provider: "ollama",
          // Default context window: Ollama's /api/tags doesn't expose context_window,
          // so 128K is a reasonable default for modern LLMs
          contextWindow: 128_000,
          // Safe default: modern models typically support tools
          supportsTools: true,
          supportsStreaming: true,
          supportsVision,
          inputCostPerMtok: 0,
          outputCostPerMtok: 0,
        };
      });
    } catch {
      return [];
    }
  }

  async healthCheck(): Promise<boolean> {
    try {
      const res = await fetch(`${this.baseUrl}/api/tags`);
      return res.ok;
    } catch {
      return false;
    }
  }
}

File: python/openharness/providers/ollama.py
"""Ollama provider — local LLM inference via Ollama API."""

from __future__ import annotations

import json
from collections.abc import AsyncIterator
from typing import Any

import httpx

from openharness.core.config import ProviderConfig
from openharness.core.events import CostUpdate, Event, TextDelta, ToolCallStart, TurnComplete
from openharness.core.exceptions import ProviderError
from openharness.core.types import (
    Message,
    ModelInfo,
    Role,
    ToolCall,
    ToolSpec,
)

from .base import BaseProvider

DEFAULT_OLLAMA_URL = "http://localhost:11434"


def _messages_to_ollama(messages: list[Message]) -> list[dict[str, Any]]:
    """Convert OpenHarness messages to Ollama chat format."""
    result: list[dict[str, Any]] = []
    for msg in messages:
        if msg.is_meta:
            continue
        entry: dict[str, Any] = {
            "role": msg.role.value if msg.role != Role.TOOL else "tool",
            "content": msg.content,
        }
        if msg.tool_calls:
            entry["tool_calls"] = [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {"name": tc.tool_name, "arguments": tc.arguments},
                }
                for tc in msg.tool_calls
            ]
        if msg.tool_results:
            # Ollama expects tool results as content in a "tool" role message
            entry["content"] = msg.tool_results[0].output if msg.tool_results else msg.content
        result.append(entry)
    return result


def _tools_to_ollama(tools: list[ToolSpec]) -> list[dict[str, Any]]:
    """Convert tool specs to Ollama tool format."""
    return [
        {
            "type": "function",
            "function": {
                "name": t.name,
                "description": t.description,
                "parameters": t.parameters,
            },
        }
        for t in tools
    ]


class OllamaProvider(BaseProvider):
    """Provider for Ollama local models."""

    def __init__(self, config: ProviderConfig) -> None:
        super().__init__(config)
        self.base_url = config.base_url or DEFAULT_OLLAMA_URL
        self._client = httpx.AsyncClient(base_url=self.base_url, timeout=300.0)

    async def complete(
        self,
        messages: list[Message],
        tools: list[ToolSpec] | None = None,
        model: str | None = None,
    ) -> Message:
        """Non-streaming completion."""
        model = model or self.config.default_model or "llama3"
        payload: dict[str, Any] = {
            "model": model,
            "messages": _messages_to_ollama(messages),
            "stream": False,
        }
        if tools:
            payload["tools"] = _tools_to_ollama(tools)

        try:
            resp = await self._client.post("/api/chat", json=payload)
            resp.raise_for_status()
        except httpx.HTTPError as exc:
            raise ProviderError(f"Ollama request failed: {exc}") from exc

        data = resp.json()
        msg_data = data.get("message", {})
        content = msg_data.get("content", "")

        # Parse tool calls if present
        tool_calls: tuple[ToolCall, ...] = ()
        if raw_calls := msg_data.get("tool_calls"):
            tool_calls = tuple(
                ToolCall(
                    id=tc.get("id", f"call_{i}"),
                    tool_name=tc["function"]["name"],
                    arguments=tc["function"].get("arguments", {}),
                )
                for i, tc in enumerate(raw_calls)
            )

        return Message(
            role=Role.ASSISTANT,
            content=content,
            tool_calls=tool_calls,
        )

    async def stream(
        self,
        messages: list[Message],
        tools: list[ToolSpec] | None = None,
        model: str | None = None,
    ) -> AsyncIterator[Event]:
        """Stream response events from Ollama."""
        model = model or self.config.default_model or "llama3"
        payload: dict[str, Any] = {
            "model": model,
            "messages": _messages_to_ollama(messages),
            "stream": True,
        }
        if tools:
            payload["tools"] = _tools_to_ollama(tools)

        total_input = 0
        total_output = 0

        try:
            async with self._client.stream("POST", "/api/chat", json=payload) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if not line.strip():
                        continue
                    try:
                        chunk = json.loads(line)
                    except json.JSONDecodeError:
                        continue

                    # Text content
                    msg = chunk.get("message", {})
                    if content := msg.get("content", ""):
                        yield TextDelta(content=content)

                    # Tool calls in streamed response
                    if raw_calls := msg.get("tool_calls"):
                        for i, tc in enumerate(raw_calls):
                            fn = tc.get("function", {})
                            yield ToolCallStart(
                                tool_name=fn.get("name", f"unknown_{i}"),
                                call_id=tc.get("id", f"call_{i}"),
                            )

                    # Done flag
                    if chunk.get("done"):
                        total_input = chunk.get("prompt_eval_count", 0)
                        total_output = chunk.get("eval_count", 0)
                        yield CostUpdate(
                            input_tokens=total_input,
                            output_tokens=total_output,
                            cost=0.0,  # Local models are free
                            model=model,
                        )

        except httpx.HTTPError as exc:
            raise ProviderError(f"Ollama streaming failed: {exc}") from exc

    def list_models(self) -> list[ModelInfo]:
        """List locally available Ollama models (synchronous for simplicity)."""
        try:
            with httpx.Client(base_url=self.base_url, timeout=10.0) as client:
                resp = client.get("/api/tags")
                resp.raise_for_status()
                data = resp.json()
        except httpx.HTTPError:
            return []

        models: list[ModelInfo] = []
        for m in data.get("models", []):
            name = m.get("name", "")
            details = m.get("details", {})
            models.append(
                ModelInfo(
                    id=name,
                    provider="ollama",
                    context_window=int(details.get("context_length", 8192)),
                    supports_tools=True,
                    supports_streaming=True,
                    supports_vision="vision" in name.lower(),
                    input_cost_per_mtok=0.0,
                    output_cost_per_mtok=0.0,
                )
            )
        return models

    async def health_check(self) -> bool:
        """Check if Ollama is running locally."""
        try:
            resp = await self._client.get("/api/tags")
            return resp.status_code == 200
        except httpx.HTTPError:
            return False

    @staticmethod
    def auto_detect() -> bool:
        """Check if Ollama is running at the default URL."""
        try:
            resp = httpx.get(f"{DEFAULT_OLLAMA_URL}/api/tags", timeout=3.0)
            return resp.status_code == 200
        except httpx.HTTPError:
            return False
